### PR TITLE
[FEATURE] simplified BinnedSpectrum implementation. Some bug fixes

### PR DESCRIPTION
**Pull request by timosachsenberg**

### CODE REVIEW

File: src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp [hunk 1]
```diff
@@ -40,91 +40,73 @@ using namespace std;
...
 
-  void BinnedSpectrum::setBinning()
+  void BinnedSpectrum::binSpectrum_(const PeakSpectrum& ps)
   {
-    if (raw_spec_.empty())
-    {
-      throw BinnedSpectrum::NoSpectrumIntegrated(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
-    }
-    bins_.clear();
+    // TODO: add assertion to check if spectrum ps is sorted by position or retrieve maximum differently
```
Comment (hroest) [inline src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp]: maybe use` OPENMS_PRECONDITION` ?

File: src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp [hunk 2]
```diff
@@ -40,91 +40,73 @@ using namespace std;
...
+      // add peak to neighboring bins
       for (Size j = 0; j < bin_spread_; ++j)
       {
-        bins_[bin_number + j + 1] = bins_.at(bin_number + j + 1) + raw_spec_[i].getIntensity();
-        // we are not in one of the first bins (0 to bin_spread)
-        //not working:  if (bin_number-j-1 >= 0)
-        if (bin_number >= j + 1)
+        bins_[idx + j + 1] = bins_.at(idx + j + 1) + p.getIntensity();
+        // prevent spreading over left boundaries
+        if (idx >= j + 1)
```
Comment (hroest) [inline src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp]: why not just write `idx-j-1 >= 0`? 

also, why dont we check to go over the right boundary? what if `idx+j+1` is larger than the vector?

**Comment by hroest** [comment]
looks pretty good, the windows test failure is unrelated

File: src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp (hunk 2 above)
Comment (timosachsenberg) [reply src/openms/source/COMPARISON/SPECTRA/BinnedSpectrum.cpp]: good point .... its copy & pasted code from the old implementation

**Comment by timosachsenberg** [comment]
To summarize what I did:
Previously BInnedSpectrum stored the full (non-binned) raw spectrum and, thus, was able to recalculate bins on demand. So, one could say it was a binned view on a raw spectrum that had to cache the precalculated view data.
Now it is a simple implementation of a binned spectrum that is created once from a raw spectrum.
I tihnk this implementation is much better and simpler and could, in principle, also be used to reimplement aview (if one ever wanted to have something like that ;) ).

**Comment by timosachsenberg** [comment]
will make some minor renames so please don't merge yet

**Comment by timosachsenberg** [comment]
ok I added operator+ to SparseVectors so in principle we should now be able to do a simple accumulation of binned spectra

**Comment by hroest** [comment]
can you check the style test:

```
The following tests FAILED:

	1750 - src/topp/SpecLibSearcher.cpp_cpplint_test (Failed)
```

File: src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp [hunk 3]
```diff
@@ -93,7 +93,10 @@ START_SECTION((double operator()(const BinnedSpectrum &spec1, const BinnedSpectr
   BinnedSpectrum bs2 (1.5,2,s2);
 
   double score = (*ptr)(bs1, bs2);
-  TEST_REAL_SIMILAR(score, 0.997576)
+  TEST_REAL_SIMILAR(score, 0.99707)
```
Comment (hroest) [inline src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp]: any reason why the numerics change?

**Comment by hroest** [comment]
how come the tests pass before the contrib PR is merged? https://github.com/OpenMS/contrib/pull/55

File: src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp (hunk 3 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp]: the reason is that the score normalization is performed by dividing by (sum1 + sum2).
Previously, sum1 and sum2 were calculated only on a subset of the dimensions (the minimum of the highest non-zero indices of both spectra). It is now calculated for the full spectrum.
I looked into the original publication and, as I understood it, the way I am doing it now is correct.

**Comment by timosachsenberg** [comment]
let's wait till travis-ci finishes. I think it will not pass the tests.
@jpfeuffer why is the windows ci passing?

File: src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp (hunk 3 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/BinnedSumAgreeingIntensities_test.cpp]: in other words: if s1 had the last non-zero coefficient at, say, index 100 and s2 at index 200, then the sums were previously only calculated up to index 100. Now it is calculated for the full range of each spectrum.

**Comment by timosachsenberg** [comment]
By the way, I also removed the check for matching precursor m/z. In my view it doesn't belong into the scoring class. There are different scenarios like matching without considering precursor m/z (m/z agnostic). Matching while only considering the precursor mass (charge agnostic). Matching with relative / absolute tolerances (normal case), etc.. I would leave this decision to the caller - otherwise, it will restrict how the scoring functions can be used.

**Comment by jpfeuffer** [comment]
There were two potential problems:
- The GitHub API changed and we have to upgrade the plugin.
- The commit status build step was not the last one.

**Comment by jpfeuffer** [comment]
But the build definitely also failed on Windows.

**Comment by timosachsenberg** [comment]
Now supports also ppm bin sizes...

**Comment by mwalzer** [comment]
cool, ppm bins sounds awesome :+1:

**Comment by timosachsenberg** [comment]
windows and mac ci seem to be related to not finding qt4

**Comment by timosachsenberg** [comment]
python Bindings need to be updated so please don't merge

**Comment by jpfeuffer** [comment]
If you reset the submodule to develop we can build again.
I would like maracluster to be introduced with the maracluster PR.
And seems like you tried to update the submodule from an even older state (without crux)

**Comment by jpfeuffer** [comment]
Windows fails. Probably missing dll_export for your functor or something like that.

**Comment by timosachsenberg** [comment]
rebuild win

**Comment by timosachsenberg** [comment]
builds on my computer and now passes on windows. travis timed out

**Comment by jpfeuffer** [comment]
Are you on Linux? Do you think travis will work for other PRs if we merge?

**Comment by timosachsenberg** [comment]
yes I am on linux. I have no idea but I could squash and merge the commits so we can easily revert it.

**Comment by jpfeuffer** [comment]
Win still works. Contrib seems to be at correct HEAD. Since travis is probably just a caching timeout, it would be fine for me to merge.

**Comment by hroest** [comment]
@timosachsenberg @jpfeuffer please do not merge branches when travis is not green. There is usually a reason for this, e.g. the build fails. Better to re-start travis and try on your local machine to be sure before merge (or you can quickly ping me).

**Pull request merged**